[android][cmake] Remove NO_EXPORT for libtorch mobile build

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40176 [android] gradle version update
* #39587 [android] test_app example linking to pytorch_android aar content
* #39588 [android] ANDROID_STL=c++_shared
* #39507 [android] Add libtorch headers to pytorch_android aar
* **#39584 [android][cmake] Remove NO_EXPORT for libtorch mobile build**

Removing `-DNO_EXPORT` for not-custom-build to be able to link to C10/A10 api.
Custom build stays the same as its main goal is to have minimum binary size, while export api functions will increase it.

Additional changes:

1. aten/src/ATen/DynamicLibrary.cpp uses libdl, if we need this functionality we will need to link result with libdl, but currently disabling this functionality for mobile.

Differential Revision: [D22111600](https://our.internmc.facebook.com/intern/diff/D22111600)

IvanKobzarev committed:
[android][cmake] Remove NO_EXPORT for libtorch mobile build

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"




[ghstack-poisoned]

ljk53 commented on c10/macros/Macros.h:
```diff
@@ -247,11 +247,19 @@ __host__ __device__
 #endif // NDEBUG
+#if defined(__ANDROID__)
+#include <assert.h>
+#define CUDA_KERNEL_ASSERT(cond)                    \
```
why do we need touch this macro? isn't it only used for CUDA?

ljk53 replied:
I see what you are trying to fix now. Isn't #40151 simpler fix? We already skip this for `__APPLE__` and I don't know why we need keep CUDA assert for mobile builds at all...

IvanKobzarev replied:
I think yes, #40151 will be simpler.
Though my understanding is that CUDA_KERNEL_ASSERT is uniform assert that is used for both - cuda and CPU in the places where sources can be executed on both of them. In this case if we disable this assert for android - we will not see this assert in debug android builds. 
I do not know what is a fallback in this case - just no-op or throwing an exception?

ljk53 replied:
I did a code search and checked the original PR #31276 - this macro was mostly used in .cuh / .cu source files. The only '.h' file using it is in `c10/util/TypeCast.h` - probably this header is shared with cuda and cpu sources?

I incline towards taking simpler fix because:
1) this macro should not be used at cpu-only or mobile-only sources, as otherwise it should use TORCH_CHECK instead; between cuda build and mobile debug build, the former should have much higher usage (android debug build should be broken for a while without being noticed due to this, right?) thus a bug is less likely to first hit android debug build without being captured by some cuda or server cpu build/test.
2) the same argument can apply to `__APPLE__` and `__HIP_PLATFORM_HCC__` - we should fix them too if this is important :>

IvanKobzarev replied:
I agree with simpler fix, I will rebase this stack once fix landed.

ljk53 replied:
And if you still think this should be fixed on Android, I'd recommend writing this `#ifdef __ANDROID__` at the same level as `__APPLE__` and `_MSC_VER` for better readability (it's already quite complicated):
```
#if defined(__APPLE__) || defined(__HIP_PLATFORM_HCC__) ...

#elif defined(__ANDROID) ...

#elif defined(_MSC_VER) ...

#else  // other platforms

#endif
```

It shouldn't share the same `NDEBUG` section in the last else branch as the `__assert_fail` will not be used.

ljk53 replied:
> I agree with simpler fix, I will rebase this stack once fix landed.

oops - I replied before seeing your comment :)

IvanKobzarev replied:
Thanks, I  updated PR with removin changes in Macros.h

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"


Removing `-DNO_EXPORT` for not-custom-build to be able to link to C10/A10 api.
Custom build stays the same as its main goal is to have minimum binary size, while export api functions will increase it.

Additional changes:
1. (c10/macros/Macros.h) asserts for debug android build
CUDA_KERNEL_ASSERT is an uniform assert that works for CPU/CUDA, it is no-op for release build, for Debug build, CPU it uses `__assert_fail`.
android_ndk libc does not have `__assert_fail`, but has `__assert2`  with the same semantic:
https://chromium.googlesource.com/android_ndk/+/refs/heads/next/platforms/android-24/arch-arm64/usr/include/assert.h

2. aten/src/ATen/DynamicLibrary.cpp uses libdl, if we need this functionality we will need to link result with libdl, but currently disabling this functionality for mobile.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"


Removing `-DNO_EXPORT` for not-custom-build to be able to link to C10/A10 api.
Custom build stays the same as its main goal is to have minimum binary size, while export api functions will increase it.

Additional changes:

1. aten/src/ATen/DynamicLibrary.cpp uses libdl, if we need this functionality we will need to link result with libdl, but currently disabling this functionality for mobile.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][cmake] Remove NO_EXPORT for libtorch mobile build"


Removing `-DNO_EXPORT` for not-custom-build to be able to link to C10/A10 api.
Custom build stays the same as its main goal is to have minimum binary size, while export api functions will increase it.

Additional changes:

1. aten/src/ATen/DynamicLibrary.cpp uses libdl, if we need this functionality we will need to link result with libdl, but currently disabling this functionality for mobile.

Differential Revision: [D22111600](https://our.internmc.facebook.com/intern/diff/D22111600)

[ghstack-poisoned]
